utils/events.ts

```diff
@@ -0,0 +1,103 @@
+import { isBefore, isAfter } from 'date-fns'
+import { toDate } from 'date-fns-tz'
+import { formatUTC, weeksBetween } from './date'
+import { ApiEvent } from 'apiClient'
+
+export type WeeklyData = {
+  week: number
+  date: Date
+  prior: Date
+  events: ApiEvent[]
+}
+
+export const withinBounds = (start: Date, end: Date) => (e: ApiEvent) => {
+  const time = toDate(e.occurred_at)
+  return isAfter(time, start) && isBefore(time, end)
+}
+export const eventsBetween = (
+  start: Date,
+  end: Date,
+  events: ApiEvent[]
+): ApiEvent[] => events.filter(withinBounds(start, end))
+
+export function uniqueById(x: readonly ApiEvent[]) {
+  return x.reduce((agg: ApiEvent[], i: ApiEvent) => {
+    if (!agg.map(({ id }) => id).includes(i.id)) {
+      return agg.concat(i)
+    }
+    return agg
+  }, [])
+}
+
+export const getAllVisibleEvents = (weeklyData: WeeklyData[]) =>
+  weeklyData.reduce((agg: ApiEvent[], { events }) => {
+    return agg.concat(events)
+  }, [])
+
+export const findMissingEvents = (
+  allEvents: readonly ApiEvent[],
+  weeklyData: WeeklyData[]
+) =>
+  allEvents.reduce((agg: ApiEvent[], x: ApiEvent) => {
+    const onlyMatched = getAllVisibleEvents(weeklyData)
+    if (onlyMatched.map(({ id }: { id: number }) => id).includes(x.id)) {
```

Comment:
The construction of the `onlyMatched` map can just be done once before the `reduce`. You could just have `getAllVisibleEvents` return a map or a set to make the existence check easier too.

Reply:
Cleaned it up a bunch and refactored to use `Map` internally.